mailbox/jpa/src/main/java/org/apache/james/mailbox/jpa/mail/JPAMessageMapper.java

```diff
@@ -285,15 +240,71 @@ public MessageMetaData move(Mailbox mailbox, MailboxMessage original) throws Mai
     }
 
     @Override
-    protected MessageMetaData copy(Mailbox mailbox, MessageUid uid, long modSeq, MailboxMessage original)
+    public MessageMetaData add(Mailbox mailbox, MailboxMessage message) throws MailboxException {
+        messageMetadataMapper.enrichMessage(mailbox, message);
+
+        return save(mailbox, message);
+    }
+
+    @Override
+    public Iterator<UpdatedFlags> updateFlags(Mailbox mailbox, FlagsUpdateCalculator flagsUpdateCalculator,
```

Comment:
the inheritance change brings this copy/paste if I understand correctly.
That's the trap of using inheritance where we should use composition.
Please go one step further by extracting pieces of code into classes that we can compose.